packager/media/formats/mp4/low_latency_segment_segmenter.cc

```diff
@@ -0,0 +1,235 @@
+      num_segments_(0) {
+  // Use the same brands for styp as ftyp.
+  styp_->major_brand = Segmenter::ftyp()->major_brand;
+  styp_->compatible_brands = Segmenter::ftyp()->compatible_brands;
+  // Replace 'cmfc' with 'cmfs' for CMAF segments compatibility.
+  std::replace(styp_->compatible_brands.begin(), styp_->compatible_brands.end(),
+               FOURCC_cmfc, FOURCC_cmfs);
+}
+
+LowLatencySegmentSegmenter::~LowLatencySegmentSegmenter() {}
+
+bool LowLatencySegmentSegmenter::GetInitRange(size_t* offset, size_t* size) {
+  VLOG(1) << "LowLatencySegmentSegmenter outputs init segment: "
+          << options().output_file_name;
+  return false;
+}
+
+bool LowLatencySegmentSegmenter::GetIndexRange(size_t* offset, size_t* size) {
+  VLOG(1) << "LowLatencySegmentSegmenter does not have index range.";
+  return false;
+}
+
+std::vector<Range> LowLatencySegmentSegmenter::GetSegmentRanges() {
+  VLOG(1) << "LowLatencySegmentSegmenter does not have media segment ranges.";
+  return std::vector<Range>();
+}
+
+Status LowLatencySegmentSegmenter::DoInitialize() {
+  return WriteInitSegment();
+}
+
+Status LowLatencySegmentSegmenter::DoFinalize() {
+  // Update init segment with media duration set.
+  RETURN_IF_ERROR(WriteInitSegment());
+  SetComplete();
+  return Status::OK;
+}
+
+Status LowLatencySegmentSegmenter::DoFinalizeSegment() {
+  return FinalizeSegment();
+}
+
+Status LowLatencySegmentSegmenter::DoFinalizeChunk() {
+  if (is_initial_chunk_in_seg_) {
+    return WriteInitialChunk();
+  }
+  return WriteChunk();
+}
+
+Status LowLatencySegmentSegmenter::WriteInitSegment() {
+  DCHECK(ftyp());
+  DCHECK(moov());
+  // Generate the output file with init segment.
+  std::unique_ptr<File, FileCloser> file(
+      File::Open(options().output_file_name.c_str(), "w"));
+  if (!file) {
+    return Status(error::FILE_FAILURE,
+                  "Cannot open file for write " + options().output_file_name);
+  }
+  std::unique_ptr<BufferWriter> buffer(new BufferWriter);
+  ftyp()->Write(buffer.get());
+  moov()->Write(buffer.get());
+  return buffer->WriteToFile(file.get());
+}
+
+Status LowLatencySegmentSegmenter::WriteInitialChunk() {
+  DCHECK(sidx());
+  DCHECK(fragment_buffer());
+  DCHECK(styp_);
+  
+  DCHECK(!sidx()->references.empty());
+  // earliest_presentation_time is the earliest presentation time of any access
+  // unit in the reference stream in the first subsegment.
+  sidx()->earliest_presentation_time =
+      sidx()->references[0].earliest_presentation_time;
+
+  if (options().segment_template.empty()) {
+    // Append the segment to output file if segment template is not specified.
+    file_name_ = options().output_file_name.c_str();
+  } else {
+    file_name_ = GetSegmentName(options().segment_template,
+                               sidx()->earliest_presentation_time,
+                               num_segments_, options().bandwidth);
+  }
+
+  // Create the segment file
+  segment_file_ = File::Open(file_name_.c_str(), "a");
```

Comment:
How does append mode work on an HTTP "file"?

Reply:
To "append" data to the the HTTP file, we aren't using the curl append mode. Rather, the file is being created and held open for the duration of 1 segment. After a chunk is created, it is immediately written to the already open file. The chunk data is stored in the  `IoCache upload_cache_` and on `IoCache::Write(...)`, a `base::WaitableEvent write_event_` is signaled, which triggers the transmission of the data to the remote file. I explored using [`CURLOPT_APPEND`](https://curl.se/libcurl/c/CURLOPT_APPEND.html), but it is only compatible with FTP. What are your thoughts? :) You have much more expertise in this area than I do.